src/hotspot/share/opto/divnode.cpp

```diff
@@ -1257,6 +1386,152 @@ const Type* ModLNode::Value(PhaseGVN* phase) const {
   return TypeLong::make( i1->get_con() % i2->get_con() );
 }
 
+//=============================================================================
+//------------------------------Idealize---------------------------------------
+Node* UModINode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  // Check for dead control input
+  if(in(0) && remove_dead_region(phase, can_reshape)) {
+    return this;
+  }
+  // Don't bother trying to transform a dead node
+  if(in(0) && in(0)->is_top()) {
+    return nullptr;
+  }
+
+  // Get the modulus
+  const Type* t = phase->type(in(2));
+  if(t == Type::TOP) {
+    return nullptr;
+  }
+  const TypeInt* ti = t->is_int();
+
+  // Check for useless control input
+  // Check for excluding mod-zero case
+  if (in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    set_req(0, nullptr);        // Yank control input
+    return this;
+  }
+
+  if(!ti->is_con()) {
+    return nullptr;
+  }
+  juint con = ti->get_con();
+  const Type* u = phase->type(in(1));
```

Comment:
This is a constant foldable bailout? Why do you do it earlier here?

Generally, I'm starting to wonder if all this code duplication makes sense in all the `Ideal` methods?

Reply:
Yes we bailout to avoid having to do all the calculations. Let me think how to avoid all these duplications.